{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7b719fb5-c640-4dcc-a94e-f4971978e034",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97c83738-40d5-4997-9ba1-51425014e46a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Paragraphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "de4cf57f-77b8-419b-9afb-a48d85227add",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of methods paragraphs: 44853\n",
      "total number of methods sentences: 641892\n"
     ]
    }
   ],
   "source": [
    "# open the paragraphs file if you don't want to run the above again\n",
    "\n",
    "filename = r\"C:\\Users\\conix\\Documents\\Corpus\\methods_corpus\\methods_paragraph_corpus.pickle\"\n",
    "with open(filename, \"rb\") as f:\n",
    "    methods_paras = pickle.load(f)\n",
    "\n",
    "\n",
    "filename = r\"C:\\Users\\conix\\Documents\\Corpus\\methods_sentencesrawtext.pickle\"        \n",
    "with open(filename, \"rb\") as f:\n",
    "    methods_sentences = pickle.load(f)\n",
    "\n",
    "print(f'total number of methods paragraphs: {len(methods_paras)}')\n",
    "print(f'total number of methods sentences: {len(methods_sentences)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8baad22-945a-4911-8f47-2eccee332843",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Convenience functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "18a183f5-0d86-40e8-b23f-47b26b812998",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to create a list of the columns from the dictionary with the classification\n",
    "def extract_strings(d):\n",
    "    strings = []\n",
    "    if isinstance(d, dict):\n",
    "        for key, value in d.items():\n",
    "            strings.append(key)\n",
    "            strings.extend(extract_strings(value))\n",
    "    elif isinstance(d, list):\n",
    "        for item in d:\n",
    "            strings.extend(extract_strings(item))\n",
    "    elif isinstance(d, str):\n",
    "        strings.append(d)\n",
    "    return strings\n",
    "\n",
    "# read json file with the annotations\n",
    "def read_jsonfile(filepath):\n",
    "    data = []\n",
    "    try:\n",
    "        with open(filepath, 'r', encoding='utf-8') as file:\n",
    "            for line in file:\n",
    "                try:\n",
    "                    data.append(json.loads(line))\n",
    "                except json.JSONDecodeError:\n",
    "                    print(f\"Error decoding line: {line}\")\n",
    "    except FileNotFoundError:\n",
    "        print(f\"The file {filepath} was not found.\")\n",
    "    return data\n",
    "\n",
    "# Function to get the labels from the jsonl\n",
    "def extract_terms_after_colons(data):\n",
    "    terms = []\n",
    "    \n",
    "    def traverse_dict(d):\n",
    "        for key, value in d.items():\n",
    "            if isinstance(value, dict):\n",
    "                traverse_dict(value)  # Recursively traverse if value is a dict\n",
    "            else:\n",
    "                # Extract the part after ':::'\n",
    "                if isinstance(value, str) and ':::' in value:\n",
    "                    terms.append(value.split(':::')[-1])\n",
    "    \n",
    "    traverse_dict(data)\n",
    "    return terms\n",
    "\n",
    "# Function to update the df for higher categories (in place)\n",
    "def recursive_update(df, classif):\n",
    "    # Go through each item in the classification\n",
    "    for key, value in classif.items():\n",
    "        if isinstance(value, list):\n",
    "            # First process subcategories (go deeper into the hierarchy)\n",
    "            for v in value:\n",
    "                if isinstance(v, dict):\n",
    "                    # Recursively update subcategories first\n",
    "                    recursive_update(df, v)\n",
    "            \n",
    "            # After processing subcategories, update the current category\n",
    "            sub_categories = [v if isinstance(v, str) else list(v.keys())[0] for v in value]\n",
    "            df.loc[df[sub_categories].eq(1).any(axis=1), key] = 1\n",
    "    \n",
    "    # Return the DataFrame to allow for chaining if needed\n",
    "    return df\n",
    "\n",
    "# Function to turn the jsonl into a complete df\n",
    "def json_to_df(data, classif):\n",
    "    # Get the columns from the classification\n",
    "    all_strings = extract_strings(classif)\n",
    "    extra_columns = [\"id\", \"displayed_text\"]\n",
    "    columns = all_strings + extra_columns\n",
    "\n",
    "    # Create a list to hold all rows\n",
    "    rows = []\n",
    "\n",
    "    # Loop over the data and create a new row for each entry\n",
    "    for i in data:\n",
    "        row = {}  # Dictionary for the current row\n",
    "\n",
    "        # Extract labels\n",
    "        labels = extract_terms_after_colons(i['label_annotations'])\n",
    "\n",
    "        # Set values for all columns in all_strings\n",
    "        for cat in all_strings:\n",
    "            row[cat] = 1 if cat in labels else 0\n",
    "\n",
    "        # Add extra columns\n",
    "        row[\"displayed_text\"] = i[\"displayed_text\"]\n",
    "        row['id'] = i['id']\n",
    "\n",
    "        # Append the row to the list of rows\n",
    "        rows.append(row)\n",
    "    \n",
    "    # Create DataFrame\n",
    "    df = pd.DataFrame(rows, columns=columns)\n",
    "\n",
    "\n",
    "\n",
    "    # Update higher categories using recursive update (ensure in-place modification)\n",
    "    df = recursive_update(df, classif)\n",
    "\n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8aadddbd-a347-4766-9f83-333f6ded983c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Our classification"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e899171-13aa-4b02-8642-30713c5804d9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "91d33b16-e925-4eb5-91e1-b503c970f283",
   "metadata": {},
   "outputs": [],
   "source": [
    "# terms and classification used for the first batch of annotations\n",
    "# for the updated terms, see classifbelow\n",
    "\n",
    "classif_old = {\n",
    "    \"Phenotype\": [\n",
    "        {\n",
    "            \"Phen datatypes\": [\n",
    "                {\"MORPH\": \n",
    "               [ \"quant morph\",\"Interbreeding_morph\",\n",
    "                {\n",
    "                    \"qual morph\": [\n",
    "                        \"color_pattern\", \n",
    "                        \"Shape\", \n",
    "                        \"Texture\", \n",
    "                        \"Ultrastructural\", \n",
    "                        \n",
    "                    ]\n",
    "                }]},\n",
    "                {\n",
    "                    \"BEHAV\": [\n",
    "                        \"Acoustic data\", \n",
    "                        \"feeding\", \n",
    "                        \"Mating behavior\"\n",
    "                    ]\n",
    "                },\n",
    "                \"ECOLOGY\"\n",
    "                            ]\n",
    "        },\n",
    "        {\n",
    "            \"Phen processing\": [\n",
    "                \"IMAGING\",\n",
    "                \"SAMPLING\",\n",
    "                \"STORAGE\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"Phen analysis\": [\n",
    "                \"phen_regression\",\n",
    "                \"phen_pylo\"\n",
    "            ]\n",
    "        }\n",
    "    ],\n",
    "    \"Genotype\": [\n",
    "        {\n",
    "            \"genot datatypes\": [\n",
    "                \"Nuclear DNA\",\n",
    "                \"Organellar DNA\",\n",
    "                \"Transcriptomic data\",\n",
    "                \"Proteomic data\",\n",
    "                \"Microsatellites\",\n",
    "                \"Whole genomes\",\n",
    "                \"Exomes\",\n",
    "                \"Genome-wide studies/SNPs\",\n",
    "                \"Epigenetic data\",\n",
    "                \"eDNA\",\n",
    "                {\n",
    "                    \"BIOCHEM\": [\n",
    "                        \"Chemotax\", \n",
    "                        \"Cytotax\"\n",
    "                    ]\n",
    "                }\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"gen processing\": [\n",
    "                {\"SEQUENCING\": [\"gen1\", \n",
    "                \"gen2\", \n",
    "                \"gen3\",]},\n",
    "                \"other\"\n",
    "                \n",
    "                \n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"gen analysis\": [\n",
    "                {\n",
    "                    \"GEN_NON_PHYLO\": [\n",
    "                        \"Distance\", \n",
    "                        \"haplowebs\", \n",
    "                        \"Fixed alt character states\", \n",
    "                        \"Clustering\", \n",
    "                        \"fen_Interbreeding\"\n",
    "                    ]\n",
    "                },\n",
    "                 \"PHYLO_SD\",\n",
    "                    \n",
    "                {\n",
    "                    \"PHYLO_TREE\": [\n",
    "                        \"Distance_based\", \n",
    "                        \"Character_based\",\n",
    "                         \"Consensus_supertree\",\n",
    "                    ]\n",
    "                },\n",
    "               \n",
    "                \"Other\",\n",
    "                \"ML_methods\"\n",
    "            ]\n",
    "        }\n",
    "    ],\n",
    "    \"Singletons\": [\n",
    "        \"Interbreeding\", \n",
    "        \"spec justification\", \n",
    "        \"Phylogenetic\", \n",
    "        \"Specimen storage location\", \n",
    "        \"sampling location\", \n",
    "        \"abbreviations & terms\", \n",
    "        \"nomenclature & history\",\n",
    "        \"BIOGEO\"\n",
    "    ]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c49c7f7-9bf0-4cc4-ac01-56d8f446ea3a",
   "metadata": {},
   "source": [
    "## New"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64ce61ae-701e-426f-82d0-3a57cdbd95b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "classif = {\n",
    "    \"PHENOTYPE\": [\n",
    "        {\n",
    "            \"Phen_data\": [\n",
    "                {\"MORPH\": \n",
    "               [ \"quant_morph\",\"interbr_morph\",\n",
    "                {\n",
    "                    \"qual_morph\": [\n",
    "                        \"color_pattern\", \n",
    "                        \"shape\", \n",
    "                        \"texture\", \n",
    "                        \"ultrastruct\", \n",
    "                        \n",
    "                    ]\n",
    "                }]},\n",
    "                {\n",
    "                    \"BEHAV\": [\n",
    "                        \"acoustic\", \n",
    "                        \"feeding\", \n",
    "                        \"mating\"\n",
    "                    ]\n",
    "                },\n",
    "                \"ECOLOGY\"\n",
    "                            ]\n",
    "        },\n",
    "        {\n",
    "            \"Phen_proc\": [\n",
    "                \"IMAGING\",\n",
    "                \"SAMPLING\",\n",
    "                \"STORAGE\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"Phen_analysis\": [\n",
    "                \"phen_nonphylo\",\n",
    "                \"phen_pylo\"\n",
    "            ]\n",
    "        }\n",
    "    ],\n",
    "    \"GENOTYPE\": [\n",
    "        {\n",
    "            \"Gen_data\": [\n",
    "                \"nuclear\",\n",
    "                \"organellar\",\n",
    "                \"transcriptomic\",\n",
    "                \"proteomic\",\n",
    "                \"tandem_repeats\",\n",
    "                \"whole_genomes\",\n",
    "                \"exomes\",\n",
    "                \"genome_wide\",\n",
    "                \"epigenetic\",\n",
    "                \"eDNA\",\n",
    "                {\n",
    "                    \"BIOCHEM\": [\n",
    "                        \"chemotax\", \n",
    "                        \"cytotax\"\n",
    "                    ]\n",
    "                }\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"Gen_proc\": [\n",
    "                {\"SEQUENCING\": [\"gen1\", \n",
    "                \"gen2\", \n",
    "                \"gen3\"]},\n",
    "                \"genproc_other\"\n",
    "                \n",
    "                \n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"Gen_analysis\": [\n",
    "                {\n",
    "                    \"GEN_NON_PHYLO\": [\n",
    "                        \"distance\", \n",
    "                        \"haplowebs\", \n",
    "                        \"fixed_alt_states\", \n",
    "                        \"clustering\", \n",
    "                        \"gen_interbr\"\n",
    "                    ]\n",
    "                },\n",
    "                 \"PHYLO_SD\",\n",
    "                    \n",
    "                {\n",
    "                    \"PHYLO_TREE\": [\n",
    "                        \"distance_based\", \n",
    "                        \"character_based\",\n",
    "                         \"consensus_supertree\",\n",
    "                    ]\n",
    "                },\n",
    "                \"MACHINE_LEARNING\"\n",
    "            ]\n",
    "        }\n",
    "    ],\n",
    "    \"Singletons\": [\n",
    "        \"interbreeding\", \n",
    "        \"rank_just\", \n",
    "        \"phylogenetic\", \n",
    "        \"specimen_storage_loc\", \n",
    "        \"sampling_loc\", \n",
    "        \"abbrev_terms\", \n",
    "        \"nomenclat_history\",\n",
    "        \"biogeo\"\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc7046d0-84ba-4c6a-93a1-96cf8e009eb9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Annotate Data\n",
    "\n",
    "We annotate in potato, using the immigration framing template with an adapted config file (implementing our classification)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b58d6957-1321-46aa-aca2-2173614b0825",
   "metadata": {},
   "source": [
    "## load newly annotated data from jsonl\n",
    "\n",
    "This process includes reading the data and making sure that higher level categories are \"1\" if their leaf categories are \"1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a44d39-70b1-4410-8bb2-b3ed7e9ffaff",
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath2 = r\"C:\\Users\\conix\\immigration_framing\\annotation_output\\stijn_conix@hotmail.com\\annotated_instances.jsonl\"\n",
    "data2 = read_jsonfile(filepath2)\n",
    "df2 = json_to_df(data2, classif)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ec42497-c705-466e-996f-459580c11308",
   "metadata": {},
   "source": [
    "## load newly annotated data from tsv\n",
    "\n",
    "This process includes reading the data and making sure that higher level categories are \"1\" if their leaf categories are \"1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bd15b4d-5ea6-455d-86de-dd2e530c168d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.read_csv(filepath1, sep = '\\t')\n",
    "\n",
    "terms = extract_strings(classif)\n",
    "missing = [j for j in extract_strings(classif) if j not in [i.split(\":::\")[1] for i in a[3:]]]\n",
    "\n",
    "df1.columns = ['user', 'id', 'displayed_text'] + [i.split(\":::\")[1] for i in a[3:]]\n",
    "df1[missing] = np.nan\n",
    "df1[terms] = df1[terms].notna().astype(int)\n",
    "df1 = recursive_update(df1, classif)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d010d6a-a2cc-4f25-b7a8-7678333abe2e",
   "metadata": {},
   "source": [
    "## Compare and fix annotations by multiple annotators"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "3a135ac2-449f-476c-b3b9-3cdf74e1415c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "86"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first make sure that the column names and shapes are identical\n",
    "# get an idea of how much disagreement there is\n",
    "df1.compare(dfx, result_names=('stijn','marlies')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd487ad0-36f5-4c23-8def-72b04847d5db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first make sure that the column names and shapes are identical\n",
    "\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "# the row you want to fix\n",
    "r = 86\n",
    "\n",
    "# get the id\n",
    "idx = shared.iloc[r,:]['id']\n",
    "\n",
    "dfx.iloc[r,:].compare(df1.iloc[r,:], result_names=('stijn','marlies'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2195bf6-7bbf-4769-80c4-859c18b210c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the text\n",
    "shared.iloc[r,:]['displayed_text']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "id": "3d3142f2-3376-49ea-abc0-934d891d9ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# choose one df (e.g. dfx) to fix in place\n",
    "# change various column row values in place\n",
    "\n",
    "dfx.loc[dfx.id == idx, 'gen processing'] = 0\n",
    "dfx.loc[dfx.id == idx, 'Specimen storage location'] = 1\n",
    "dfx.loc[dfx.id == idx, 'Genotype'] = 0\n",
    "\n",
    "\n",
    "# use this to save the corrected df each time you make changes\n",
    "dfx.to_csv(r'C:\\Users\\conix\\Dropbox\\FNRS project taxonomy\\methods in taxonomy\\coding_trial\\firstTry\\corrected_df2.csv', sep = ';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "ce74f9f4-170e-4449-919d-c352a3944d2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "178    0\n",
       "Name: gen analysis, dtype: int64"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfx.loc[dfx.id == idx, 'gen analysis']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "affd714b-ff3b-4e51-ab5b-588b831227cc",
   "metadata": {},
   "source": [
    "# Annotated Data\n",
    "\n",
    "We want a df with a columns for every category, including all higher levels in the hierarchy, and for each sample for each column a 1 or 0."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba3f52bc-b3a0-4a26-afb0-8fa1640d41f7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Load first batch\n",
    "\n",
    "The first 100 coded samples (corrected) were done using slightly different terms (see classif_old). Update these terms here, and load the df.\n",
    "Note: 'checked_samples' are the ones that both Marlies and I did. 'corrected_dfx' are my 403 coded samples, of which Marlies checked 100."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "722cef21-ae79-4b9b-8893-66b4ddafefd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>Phen_data</th>\n",
       "      <th>MORPH</th>\n",
       "      <th>quant_morph</th>\n",
       "      <th>qual_morph</th>\n",
       "      <th>color_pattern</th>\n",
       "      <th>shape</th>\n",
       "      <th>texture</th>\n",
       "      <th>ultrastruct</th>\n",
       "      <th>...</th>\n",
       "      <th>bad sample</th>\n",
       "      <th>sampling_loc</th>\n",
       "      <th>abbrev_terms</th>\n",
       "      <th>nomenclat_history</th>\n",
       "      <th>biogeo</th>\n",
       "      <th>id</th>\n",
       "      <th>displayed_text</th>\n",
       "      <th>Gen_analysis</th>\n",
       "      <th>phen_pylo</th>\n",
       "      <th>SEQUENCING</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>./Corpus/Pensoft/zookeys-315-055.json_0</td>\n",
       "      <td>During each cruise, specimens were sorted onbo...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>92</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>./Corpus/Zootaxa/4/1/zootaxa_4181_1_1.json_1</td>\n",
       "      <td>The discovered mines were documented by photog...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>93</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>./Corpus/Zootaxa/4/7/zootaxa_4779_4_1.json_0</td>\n",
       "      <td>Imagines were reared from larvae in containers...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>./Corpus/Zootaxa/2/zootaxa_2676_1_2.json_1</td>\n",
       "      <td>Pristina jenkinae, Rhyacodrilus subterraneus, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>226</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>./Corpus/Zootaxa/3/6/zootaxa_3637_4_1.json_0</td>\n",
       "      <td>A total of 20 sites was sampled throughout the...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  PHENOTYPE  Phen_data  MORPH  quant_morph  qual_morph  \\\n",
       "0           4          1          0      0            0           0   \n",
       "1          92          1          1      1            0           1   \n",
       "2          93          1          1      1            0           1   \n",
       "3          94          1          0      0            0           0   \n",
       "4         226          1          1      1            1           0   \n",
       "\n",
       "   color_pattern  shape  texture  ultrastruct  ...  bad sample  sampling_loc  \\\n",
       "0              0      0        0            0  ...           0             0   \n",
       "1              0      1        0            0  ...           0             0   \n",
       "2              0      0        0            0  ...           0             0   \n",
       "3              0      0        0            0  ...           0             0   \n",
       "4              0      0        0            0  ...           0             1   \n",
       "\n",
       "   abbrev_terms  nomenclat_history  biogeo  \\\n",
       "0             0                  0       0   \n",
       "1             0                  0       0   \n",
       "2             0                  0       0   \n",
       "3             0                  0       0   \n",
       "4             0                  0       1   \n",
       "\n",
       "                                             id  \\\n",
       "0       ./Corpus/Pensoft/zookeys-315-055.json_0   \n",
       "1  ./Corpus/Zootaxa/4/1/zootaxa_4181_1_1.json_1   \n",
       "2  ./Corpus/Zootaxa/4/7/zootaxa_4779_4_1.json_0   \n",
       "3    ./Corpus/Zootaxa/2/zootaxa_2676_1_2.json_1   \n",
       "4  ./Corpus/Zootaxa/3/6/zootaxa_3637_4_1.json_0   \n",
       "\n",
       "                                      displayed_text  Gen_analysis  phen_pylo  \\\n",
       "0  During each cruise, specimens were sorted onbo...           NaN        NaN   \n",
       "1  The discovered mines were documented by photog...           NaN        NaN   \n",
       "2  Imagines were reared from larvae in containers...           NaN        NaN   \n",
       "3  Pristina jenkinae, Rhyacodrilus subterraneus, ...           NaN        NaN   \n",
       "4  A total of 20 sites was sampled throughout the...           NaN        NaN   \n",
       "\n",
       "   SEQUENCING  \n",
       "0           0  \n",
       "1           0  \n",
       "2           0  \n",
       "3           0  \n",
       "4           0  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first batch\n",
    "\n",
    "# 100 checked by two annotators (Marlies and Stijn)\n",
    "dfc = pd.read_csv(r'C:\\Users\\conix\\Dropbox\\FNRS project taxonomy\\methods in taxonomy\\coding_trial\\firstTry\\checked_samples.csv', sep = ';')\n",
    "\n",
    "# fix typos\n",
    "dfc = dfc.rename(columns = {'gen alaysis': \"gen analysis\", \"Phen_pylo\": 'phen_pylo'})\n",
    "\n",
    "# add new high-level category SEQUENCING\n",
    "# this combines the three generations of sequencing\n",
    "\n",
    "dfc[\"SEQUENCING\"] = np.where(((dfc['gen1'] == 1) | (dfc['gen2'] == 1) | (dfc['gen3'] == 1)), 1,0)\n",
    "\n",
    "# remove Other from the old classif\n",
    "# it was not used, and has been removed after those 403\n",
    "\n",
    "old_terms = extract_strings(classif_old)\n",
    "old_terms.remove('Other')\n",
    "\n",
    "# replace old terms by new ones\n",
    "\n",
    "replace_dict = {old_terms[i]: extract_strings(classif)[i] for i in range(len(extract_strings(classif))) }\n",
    "dfc = dfc.rename(columns=replace_dict)\n",
    "\n",
    "dfc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3fccaf89-f9ed-43ba-857b-3d1d71403e73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Phen datatypes</th>\n",
       "      <th>MORPH</th>\n",
       "      <th>quant morph</th>\n",
       "      <th>qual morph</th>\n",
       "      <th>color_pattern</th>\n",
       "      <th>Shape</th>\n",
       "      <th>Texture</th>\n",
       "      <th>Ultrastructural</th>\n",
       "      <th>...</th>\n",
       "      <th>Specimen storage location</th>\n",
       "      <th>bad sample</th>\n",
       "      <th>sampling location</th>\n",
       "      <th>abbreviations &amp; terms</th>\n",
       "      <th>nomenclature &amp; history</th>\n",
       "      <th>BIOGEO</th>\n",
       "      <th>id</th>\n",
       "      <th>displayed_text</th>\n",
       "      <th>gen alaysis</th>\n",
       "      <th>Phen_pylo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>./Corpus/EJT/10_5852_ejt_2021_735_1243.json_0</td>\n",
       "      <td>The material examined was collected in fragmen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>./Corpus/Zootaxa/1/zootaxa_1920_1_5.json_0</td>\n",
       "      <td>Invertebrate samples\\nwere collected using a h...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>./Corpus/Zootaxa/4/7/zootaxa_4729_2_8.json_0</td>\n",
       "      <td>The nymphs were collected in the stream by han...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>./Corpus/Pensoft/phytokeys-47-059.json_0</td>\n",
       "      <td>We verified both the endemic status and the di...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>./Corpus/Pensoft/zookeys-315-055.json_0</td>\n",
       "      <td>During each cruise, specimens were sorted onbo...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 74 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  Phenotype  Phen datatypes  MORPH  quant morph  qual morph  \\\n",
       "0           0          1               0      0            0           0   \n",
       "1           1          1               0      0            0           0   \n",
       "2           2          1               0      0            0           0   \n",
       "3           3          1               0      0            0           0   \n",
       "4           4          1               0      0            0           0   \n",
       "\n",
       "   color_pattern  Shape  Texture  Ultrastructural  ...  \\\n",
       "0              0      0        0                0  ...   \n",
       "1              0      0        0                0  ...   \n",
       "2              0      0        0                0  ...   \n",
       "3              0      0        0                0  ...   \n",
       "4              0      0        0                0  ...   \n",
       "\n",
       "   Specimen storage location  bad sample  sampling location  \\\n",
       "0                          0           0                  1   \n",
       "1                          1           0                  0   \n",
       "2                          0           0                  0   \n",
       "3                          0           0                  0   \n",
       "4                          0           0                  0   \n",
       "\n",
       "   abbreviations & terms  nomenclature & history  BIOGEO  \\\n",
       "0                      1                       0       0   \n",
       "1                      0                       0       0   \n",
       "2                      0                       0       0   \n",
       "3                      0                       0       0   \n",
       "4                      0                       0       0   \n",
       "\n",
       "                                              id  \\\n",
       "0  ./Corpus/EJT/10_5852_ejt_2021_735_1243.json_0   \n",
       "1     ./Corpus/Zootaxa/1/zootaxa_1920_1_5.json_0   \n",
       "2   ./Corpus/Zootaxa/4/7/zootaxa_4729_2_8.json_0   \n",
       "3       ./Corpus/Pensoft/phytokeys-47-059.json_0   \n",
       "4        ./Corpus/Pensoft/zookeys-315-055.json_0   \n",
       "\n",
       "                                      displayed_text  gen alaysis  Phen_pylo  \n",
       "0  The material examined was collected in fragmen...          NaN        NaN  \n",
       "1  Invertebrate samples\\nwere collected using a h...          NaN        NaN  \n",
       "2  The nymphs were collected in the stream by han...          NaN        NaN  \n",
       "3  We verified both the endemic status and the di...          NaN        NaN  \n",
       "4  During each cruise, specimens were sorted onbo...          NaN        NaN  \n",
       "\n",
       "[5 rows x 74 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 400, of which the same 100 checked by two annotators (Marlies and Stijn)\n",
    "dfc2 = pd.read_csv(r'C:\\Users\\conix\\Dropbox\\FNRS project taxonomy\\methods in taxonomy\\coding_trial\\firstTry\\corrected_dfx.csv', sep = ';')\n",
    "\n",
    "dfc2.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
